### ISSUE TITLE
challenge-customized-greetings

### ISSUE THREAD

**Issue by Shuchong-LI** [labex-labs/scenarios#341]
> **Please read [For Newbie](https://github.com/labex-labs/scenarios/issues/18) first. 请先阅读 [新手须知](https://github.com/labex-labs/scenarios/issues/18)。**

## Basic

> Basic information about the scenario, which can be changed if required.
> 实验/挑战的基本信息，如有需要，可以修改。

- **Suggest Title 建议标题**:Customized Greetings with Python Argparse
- **Suggest Steps 建议步骤数**:1
- **Suggest Image 建议环境**: `webide-ubuntu:2204`

## Sources

> Issues has provided the material; the part not provided must be completed by the author.
> Issues 已经提供了材料；没有提供的部分由作者自行补充完成。

- [x] Full Text 正文内容
- [x] Reference Answers 参考答案
- [x] Verify Script 检测脚本

## Skills

> Suggested skills to include can be modified as needed. **The main skill points that should be included in this section are bolded**. From [Skills Tree](https://labex-dev.feishu.cn/base/bascnNz4Nqjqgqm1Nm5AYke6xxb?table=tblV5pGIsGZMxmE9&view=vewZpkLWRp).
> 建议包括的 Skills ID，可以根据需要增删。**本章节应包含的主要技能点已加粗**，技能点来自 [技能树](https://labex-dev.feishu.cn/base/bascnNz4Nqjqgqm1Nm5AYke6xxb?table=tblV5pGIsGZMxmE9&view=vewZpkLWRp)

- **`python/argparse`**

**Comment by Shuchong-LI** [comment]
# Customized Greetings with Python Argparse

## Intro

As a software developer, you often need to write command line tools that can take in user input and perform some task based on the input. One important tool for handling command line input in Python is the `argparse` module, which provides an easy and flexible way to parse command line arguments.

In this programming challenge, you are tasked with writing a Python script that takes in a name and an optional greeting message as command line arguments, and generates a customized greeting message based on the inputs. The script should use the `argparse` module to parse the command line arguments and have proper error handling for when the `--name` argument is not provided.

This challenge will test your knowledge of Python's `argparse` module, as well as your ability to write clear and concise code with proper error handling and documentation.

## Problem Description

You need to write a Python script called `greet.py` that will take in a name and an optional greeting as command line arguments and print out a customized greeting message.

The script should take in the following arguments:

- `--name` (required): the name of the person to greet
- `--greeting` (optional): a custom greeting message to use instead of the default "Hello"

If the `--greeting` argument is not provided, the script should use the default greeting "Hello". If the `--name` argument is not provided, the script should print an error message and exit.

## Example Usage

```bash
python greet.py --name Alice
```

Output: "Hello, Alice!"

```bash
python greet.py --name Bob --greeting "Hi there"
```

Output: "Hi there, Bob!"

```bash
python greet.py --greeting "Hi there"
```

Output: "Error: name argument is required"

## Constraints

- The script should use the `argparse` module to parse command line arguments.
- The script should have proper error handling for when the `--name` argument is not provided.
- The script should have a docstring explaining its usage and inputs/outputs.

---

## Solution: Python Argparse

```python
import argparse

def greet(name: str, greeting: str = "Hello") -> str:
    """
    Generates a customized greeting message.
    
    Args:
    - name: a string representing the name of the person to greet
    - greeting: an optional string representing a custom greeting message
    
    Returns:
    - a string representing the customized greeting message
    """
    return f"{greeting}, {name}!"

def main():
    """
    Parse command line arguments and generate a customized greeting message.
    """
    parser = argparse.ArgumentParser(description='Generate a customized greeting message.')
    parser.add_argument('--name', required=True, help='The name of the person to greet')
    parser.add_argument('--greeting', help='A custom greeting message to use instead of the default "Hello"')
    args = parser.parse_args()
    
    name = args.name
    greeting = args.greeting if args.greeting else "Hello"
    
    print(greet(name, greeting))

if __name__ == '__main__':
    main()
```

## Python unittest code
    
```python
import unittest
from unittest.mock import patch
from io import StringIO
import sys
import argparse
import greet

class TestGreet(unittest.TestCase):
    def test_greet_without_optional_argument(self):
        expected_output = "Hello, Alice!\n"
        with patch('sys.stdout', new=StringIO()) as fake_out:
            sys.argv = ['greet.py', '--name', 'Alice']
            greet.main()
            self.assertEqual(fake_out.getvalue(), expected_output)

    def test_greet_with_optional_argument(self):
        expected_output = "Hi there, Bob!\n"
        with patch('sys.stdout', new=StringIO()) as fake_out:
            sys.argv = ['greet.py', '--name', 'Bob', '--greeting', 'Hi there']
            greet.main()
            self.assertEqual(fake_out.getvalue(), expected_output)

    def test_greet_without_name_argument(self):
        expected_output = "Error: the following arguments are required: --name\n"
        with patch('sys.stderr', new=StringIO()) as fake_err:
            sys.argv = ['greet.py', '--greeting', 'Hi there']
            with self.assertRaises(SystemExit):
                greet.main()
            self.assertEqual(fake_err.getvalue(), expected_output)

if __name__ == '__main__':
    unittest.main()
```

## Summary
In summary, this programming challenge tests your knowledge and skills in using `argparse` module to parse command line arguments, writing clear and concise code with proper error handling and documentation, and creating test cases to verify that the code works as expected. It provides an opportunity to practice working with command line arguments, which is a common task in many Python projects. By completing this challenge, you will have a better understanding of how to use `argparse` to create user-friendly command line interfaces for your Python scripts and how to write effective test cases to ensure the correctness of your code.

**Comment by Shuchong-LI** [comment]
1.greet.py文件需要为用户建立好，并注意搭建好代码框架。
2.verify脚本可以运行unittest code并grep是否有OK字样。
